cl/search/management/commands/cl_index_parent_and_child_docs.py

```diff
@@ -53,21 +75,155 @@ def log_last_document_indexed(
+            )
+        case EventTable.DOCKET_ENTRY:
+            oldest_unique_events = (
+                DocketEntryEvent.objects.filter(
+                    pgh_id__gte=pk_offset,
+                    pgh_created_at__gte=start_date,
+                    pgh_created_at__lte=end_date,
+                )
+                .order_by("id", "pgh_created_at")
+                .distinct("id")
+            )
+        case EventTable.RECAP_DOC:
+            oldest_unique_events = (
+                RECAPDocumentEvent.objects.filter(
+                    pgh_id__gte=pk_offset,
+                    pgh_created_at__gte=start_date,
+                    pgh_created_at__lte=end_date,
+                )
+                .order_by("id", "pgh_created_at")
+                .distinct("id")
+            )
+    return oldest_unique_events
+
+
+def get_documents_to_update_or_remove(
+    iterable: Iterable, search_type: str, event_doc_type: EventTable | None
+) -> tuple[list[int], list[tuple[int, list[str]]], list[int]]:
+    """Determines the documents to update or remove based on changes in
+    specified fields.
+
+    :param iterable: An iterable of the event table instances.
+    :param search_type: The search type related to the update/remove action.
+    :param event_doc_type: An optional EventTable enum member specifying the
+    document type to be processed.
+    :return: A four tuple containing, a list of strings containing the fields
+    to be updated, a list containing IDs of parent documents to update, a list
+    containing IDs of child documents to update, a list containing IDs of
+    documents to delete.
+    """
+
+    main_documents_to_update = []
+    child_documents_to_update = []
+    documents_to_delete = []
+    match search_type:
+        case SEARCH_TYPES.RECAP if event_doc_type == EventTable.DOCKET:
+            tracked_set = Docket.es_rd_field_tracker
+            fields_map = recap_document_field_mapping["save"][Docket][
+                "docket_entry__docket"
+            ]
+            document_model = Docket
+            related_child_name = "docket_entries"
+
+        case SEARCH_TYPES.RECAP if event_doc_type == EventTable.DOCKET_ENTRY:
+            tracked_set = DocketEntry.es_rd_field_tracker
+            fields_map = recap_document_field_mapping["save"][DocketEntry][
+                "docket_entry"
+            ]
+            document_model = DocketEntry
+            related_child_name = "recap_documents"
+
+        case SEARCH_TYPES.RECAP if event_doc_type == EventTable.RECAP_DOC:
+            tracked_set = RECAPDocument.es_rd_field_tracker
+            fields_map = recap_document_field_mapping["save"][RECAPDocument][
+                "self"
+            ]
+            document_model = RECAPDocument
+            related_child_name = ""
+
+        case _:
+            return [], [], []
+
+    for event in iterable:
+        try:
+            current_instance = document_model.objects.get(pk=event.id)
```

Comment:
This line catches my eye. Is there a way to do this that wouldn't make one request per item?

Reply:
Yes, here we have a trade-off between memory usage and database hits. Since we need to compare the fields that changed between the history table event and the current instance in the database, I believe the best approach is to load multiple items at a time. This reduces the hits to the database while also preventing excessive memory usage in case there are too many events to check. Therefore, events and current instances are retrieved in chunks defined by the `chunk_size` command argument. I've refactored it according to this strategy.